acceptance/indy_acceptance/indy-tests/test_scripts/acceptance_tests/remove_and_add_role.py

```diff
@@ -0,0 +1,421 @@
+            else:
+                if message is None:
+                    message = "Steward can create another Steward (should fail)"
+                self.steps.get_last_step().set_message(message)
+
+            # 16. Verify that a Steward cannot create a Trustee.
+            self.steps.add_step("Verify that a Steward cannot create a Trustee")
+            (temp, message) = await self.add_nym(steward1_did, trustee1_did, trustee1_verkey,
+                                                 None, Roles.TRUSTEE, error_code=304)
+
+            if temp:
+                print(Colors.OKGREEN + "::PASS::Validated that a Steward cannot create a Trustee!\n" + Colors.ENDC)
+            else:
+                if message is None:
+                    message = "Steward can create a Trustee (should fail)"
+                self.steps.get_last_step().set_message(message)
+
+            # 17. Using TrustAnchor1 to add a NYM.
+            self.steps.add_step("Using TrustAnchor1 to add a NYM")
+            await self.add_nym(trustanchor1_did, user1_did, user1_verkey, None, None)
+
+            # 18. Verify GET NYM - User1.
+            self.steps.add_step("Verify GET NYM - User1")
+            await self.get_nym(trustanchor1_did, user1_did)
+
+            # 19. Verify that TrustAnchor cannot create another TrustAnchor.
+            self.steps.add_step("Verify that TrustAnchor cannot create another TrustAnchor")
+            (temp, message) = await self.add_nym(trustanchor1_did, trustanchor2_did, trustanchor2_verkey,
+                                                 None, Roles.TRUST_ANCHOR, error_code=304)
+            if temp:
+                print(Colors.OKGREEN + "::PASS::Validated that a TrustAnchor cannot create another TrustAnchor!\n" + Colors.ENDC)
+            else:
+                if message is None:
+                    message = "TrustAnchor can create another TrustAnchor (should fail)"
+                self.steps.get_last_step().set_message(message)
+
+            # 20. Using default Trustee to remove new roles.
+            bug_is_430 = "Bug: https://jira.hyperledger.org/browse/IS-430"
+            self.steps.add_step("Using default Trustee to remove new roles")
+            message_20 = ""
+            (temp, message) = await self.add_nym(default_trustee_did, trustee1_did, trustee1_verkey,
+                                                 None, Roles.NONE)
+            result = temp
+            if not temp:
+                message_20 += "\nCannot remove Trustee1's role - " + message
+            else:
+                (temp, message) = await self.get_nym(default_trustee_did, trustee1_did)
+                if not temp:
+                    message_20 += "\nCannot check self.get_nym for Trustee1 - " + message
+                else:
+                    if not TestScenario09.check_role_in_retrieved_nym(message, Roles.NONE):
+                        temp = False
+                        message_20 += "\nCannot remove Trustee1's role"
```

Comment:
Ideally we should not continue the test in case some its step was failed. So, if any exception is raised, the test should not try to recover and continue execution (unless this exception is expected by the test). The whole test must fail at this step. Such the approach simplifies tests. It eliminates necessity to use conditional logic for steps below that depend on the step which can fail.

If in case some step has failed, we still have some steps below that may be verified, then this indicates that we actually have several branches in the test and such the branches should be divided into separate tests.

We think that in one of next pull request we should divide test scenarios in such a way that we will no need to continue a test if some step has failed. This will allow us to get rid of translating exceptions to return values in helper methods and correspondingly of using conditional logic in handling step results and performing dependent steps below. This, in turn, will significantly simplify the code of tests.

Reply:
These acceptance tests were written to match the acceptance test documents.  When the documents change, the tests will be refactored to match.  The acceptance tests documents are used to validate the release candidate.  If a single test within an acceptance test document fails, the tester does not stop all testing.